### PR TITLE
Proofs for quadratic residuosity and quadratic nonresiduosity.

### PR DESCRIPTION
**Pull request by miha-stopar**
Implementation of QR and QNR proofs as presented in famous Goldwasser, Micali, and Rackoff paper.

Signed-off-by: Miha Stopar <miha.stopar@xlab.si>

### CODE REVIEW

File: qrproofs/qnr.go [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+import (
+	"errors"
+	"github.com/xlab-si/emmy/common"
+	"github.com/xlab-si/emmy/dlog"
+	"math/big"
+)
+
+// Statistical zero-knowledge proof	of quadratic non-residousity (implemented for historical
+// reasons).
+
```
Comment (mancabizjak) [inline qrproofs/qnr.go]: Remove empty line and correct spacing in L10 comment.

File: qrproofs/qnr.go [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+	return true, nil
+}
+
+type QNRProver struct {
+	QR *dlog.QR
+	Y  *big.Int
+	w  *big.Int
+}
+
+func NewQNRProver(qr *dlog.QR, y *big.Int) *QNRProver {
```
Comment (mancabizjak) [inline qrproofs/qnr.go]: Please use this instead:

````go
return &QNRProver{
 	QR: qr,
        Y:  y,
 }
````

File: qrproofs/qnr.go [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+func (prover *QNRProver) SetProofRandomData(w *big.Int) {
+	prover.w = w
+}
+
+func (prover *QNRProver) GetProofData(challenge *big.Int) (int, error) {
+	isQR, err := prover.QR.IsQR(challenge)
+	if err != nil {
+		return 0, err
+	}
+	var typ int
```
Comment (mancabizjak) [inline qrproofs/qnr.go]: Can you please rename this to something else :slightly_smiling_face:  (for instance, the same name as in the `QRVerifier`)

File: server/qnr.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+		if err := s.send(resp, stream); err != nil {
+			return err
+		}
+
+		req, err = s.receive(stream)
+		if err != nil {
+			return err
+		}
+
+		typ := req.GetEint()
```
Comment (mancabizjak) [inline server/qnr.go]: Please use some other name (the same one as in the `QRVerifier`).

File: qrproofs/qnr.go [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+		}
+	}
+	return true
+}
+
+type QNRVerifier struct {
+	QR    *dlog.QR
+	x     *big.Int
+	y     *big.Int
+	typ   int
```
Comment (mancabizjak) [inline qrproofs/qnr.go]: I would prefer to rename this field to `verifierType` or `qrType` or something else that makes sense. I know `type` is reserved, but just `typ` kind of bothers me, sorry :smile: .
The same goes for all the other usages of `typ`.

File: qrproofs/qr.go [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+type QRProver struct {
+	DLog *dlog.ZpDLog
+	Y    *big.Int
+	y1   *big.Int
+	r    *big.Int
+}
+
+func NewQRProver(dlog *dlog.ZpDLog, y1 *big.Int) *QRProver {
+	y, _ := dlog.Multiply(y1, y1)
+	prover := QRProver{
```
Comment (mancabizjak) [inline qrproofs/qr.go]: Please replace with

```go
return &QRProver{
 	DLog: dlog,
 	Y:    y,
 	y1:   y1,
 }
```

File: qrproofs/qr.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+
+type QRVerifier struct {
+	DLog      *dlog.ZpDLog
+	x         *big.Int
+	y         *big.Int
+	challenge *big.Int
+}
+
+func NewQRVerifier(y *big.Int, dlog *dlog.ZpDLog) *QRVerifier {
+	verifier := QRVerifier{
```
Comment (mancabizjak) [inline qrproofs/qr.go]: Please replace with

````go
return &QRVerifier{
 	DLog: dlog,
 	y:    y,
 }
````

File: qrproofs/qr.go [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+	if verifier.challenge.Cmp(big.NewInt(0)) == 0 {
+		if z2.Cmp(verifier.x) == 0 {
+			return true
+		} else {
+			return false
+		}
+	} else {
+		s := new(big.Int).Mul(verifier.x, verifier.y)
+		s.Mod(s, verifier.DLog.P)
+		if z2.Cmp(s) == 0 {
```
Comment (mancabizjak) [inline qrproofs/qr.go]: Instead of L105-L110, you can just `return z2.Cmp(s) == 0`.

File: qrproofs/qr.go [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+	c := common.GetRandomInt(big.NewInt(2)) // 0 or 1
+	verifier.challenge = c
+	return c
+}
+
+func (verifier *QRVerifier) Verify(z *big.Int) bool {
+	z2 := new(big.Int).Mul(z, z)
+	z2.Mod(z2, verifier.DLog.P)
+	if verifier.challenge.Cmp(big.NewInt(0)) == 0 {
+		if z2.Cmp(verifier.x) == 0 {
```
Comment (mancabizjak) [inline qrproofs/qr.go]: Instead of L96-L101, you can `return z2.Cmp(verifier.x) == 0`

File: qrproofs/qnr.go [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+	r     *big.Int
+	pairs []*common.Pair
+}
+
+func NewQNRVerifier(qr *dlog.QR, y *big.Int) *QNRVerifier {
+	verifier := QNRVerifier{
```
Comment (mancabizjak) [inline qrproofs/qnr.go]: Use this instead:

````go
return &QNRVerifier{
	QR: qr,
	y:  y,
}
````

File: test/qr_test.go [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+				continue
+			}
+			if !isQR {
+				fmt.Println("found-------------------")
+				fmt.Println(y)
+				break
+			}
+		}
+	*/
+	y, _ := new(big.Int).SetString("12104178836609367680573806127379667907452906068454411069540554013287299560619180831355030260792398116234585889489580787593876656468666620239704236202828625502145041222925221901747074982936094534863675673392263672910937547483238701395223837362913804135100562910313510524388491518118503153440479519531614792845582743827952578371105856240886035300256188119597496494955532842085616018276731618147827132295654573847443973865791627336995666490060797108039052526091724109438973877494599020695354285996982775389058505616271009101591441286538176405813622092518363027538767195421845003207037113604997900115835538908295151715618", 10)
```
Comment (mancabizjak) [inline test/qr_test.go]: Can we avoid having this long string here? Shouldn't this be read from config instead, like the rest of test data?

**Comment by mancabizjak** [comment]
Just to recap what we talked about, I'm giving a suggestion here how we could make this PR more managable (this is actually the same thing @miha-plesko told me in #9 for my large PR).

It would be easier to review the PR if it was organized into several smaller commits, for instance:
1. Implement QR and QNR proofs  - the actual crypto part (`qrproofs` package and `dlog/qr.go`)
2. Implement communication layer for QR scheme (`client/qr.go` and `server/qr.go`, changes in `protobuf` package and translations between types in `common/types.go`)
3. Implement communication layer for QNR scheme (`client/qnr.go` and `server/qnr.go`)
4. Bind QR, QNR proofs to client and server - this would cover also tests and appropriate configuration (`client/client.go` and `server/server.go`, `test/qr_test.go`, and modifications in `config` package) 
5. Update README

No need to modify git history now, I'm only adding this for future reference :wink:

File: common/types.go [hunk 12]
```diff
@@ -32,6 +32,28 @@ func ToPbECGroupElement(el *ECGroupElement) *pb.ECGroupElement {
 	return &x
 }
 
+// Pair is the same as ECGroupElement, but to be used in non EC schemes when a pair of
```
Comment (miha-stopar) [inline common/types.go]: I was thinking about this, but due to readability I decided to go for two structs. Using only Pair is an option, but let's wait and see if some other similar duplications will be considered and decide about this then.

File: client/qnr.go [hunk 13]
```diff
@@ -0,0 +1,130 @@
...
+	}
+	if resp.GetEmpty() == nil {
+		return false, fmt.Errorf("should receive empty message")
+	}
+
+	proved := false
+	// the client has to prove for all i - if in one iteration the knowledge
+	// is not proved, the protocol is stopped
+	for i := 0; i < m; i++ {
+		msg := &pb.Message{
```
Comment (miha-stopar) [inline client/qnr.go]: Changed. Note that in Schnorr the naming is wrong - getProofRandomData actually sends proofRandomData and gets challenge (PedersenCommitment).

Comment (mancabizjak) [inline client/qnr.go]: Great! Ok, I'll take note of this and we'll make corrections for Schnorr.

File: common/types.go (hunk 12 above)
Comment (mancabizjak) [inline common/types.go]: Okay, I agree. Sounds good to me.

File: test/qr_test.go (hunk 11 above)
Comment (mancabizjak) [inline test/qr_test.go]: Maybe we should introduce a YAML file `testdata.yml` which we can put in `test/testdata` folder. In future :slightly_smiling_face:

**Pull request merged**